### PR TITLE
[CELEBORN-696] Fix bugs related with shutting down and excluded workers

### PR DESCRIPTION
**Pull request by waitinfuture**
### What changes were proposed in this pull request?
1. Foreach PartitionLocation returned from Register Shuffle, remove from worker's local excluded list to refresh the local information.
2. ChangeLocationResponse will also return whether oldPartition is excluded in LifecycleManager. If so, remove it from Executor's excluded list.
3. Always trigger commit files for shutting down workers returned from HeartbeatFromApplicationResponse.
4. HeartbeatFromWorker sends the correct disk infos regardless of shutting down or not.

After this PR, the priority of excluded list is Master > LifecycleManager > Executor.

### Why are the changes needed?
During test with graceful turned on(workers have static rpc/push/fetch/replicate ports) and consistently restart one out of three workers, I encountered several bugs.
1. First I killed worker A, then Executor's client's local excluded list will contain A, after A stopped, I started it again, then master will offer slots on A, so we should remove from the executor's excluded list then.
2. When I kill-and-start a worker twice in a short time smaller than the app heartbeat interval, the second time WorkerStatusTracker will not trigger commit files because the local cache for the worker has not been refreshed.
3. When a worker is shutting down, in its heartbeat it passes empty diskInfos, and master blindly added to excluded list. We want a worker be either in the excluded list, or in the shutting down list, exclusively. If a worker is in excluded list, then LIfecycleManager will not trigger commit files when handle heartbeat response; on the other hand, if a worker is in the shutting down list, LifecycleManager will trigger commit files on it. So we must make it correct that a shutting down worker be in the shutting down list.

### Does this PR introduce _any_ user-facing change?
Yes, it fixes several bugs described above.

### How was this patch tested?
Manual test.

### CODE REVIEW

**Comment by waitinfuture** [comment]
cc @RexXiong @AngersZhuuuu @pan3793

File: client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala [hunk 1]
```diff
@@ -198,6 +201,6 @@ class WorkerStatusTracker(
     shuttingWorkers.retainAll(newShutdownWorkers)
     val shutdownList = newShutdownWorkers.asScala.filterNot(shuttingWorkers.asScala.contains).asJava
     shuttingWorkers.addAll(newShutdownWorkers)
-    shutdownList
+    newShutdownWorkers
```
Comment (RexXiong) [inline client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:204]: If so, we can simplify this code.

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java [hunk 2]
```diff
@@ -209,7 +209,7 @@ public void updateWorkerHeartbeatMeta(
           });
     }
     appDiskUsageMetric.update(estimatedAppDiskUsage);
-    if (!blacklist.contains(worker) && disks.isEmpty()) {
+    if (!blacklist.contains(worker) && disks.isEmpty() && !shutdownWorkers.contains(worker)) {
```
[the remark this answers is not in the document]
Comment (RexXiong) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:212]: > Right, so I think it's better to alway send the true status regardless it's shutting down or not, WDYT?
> 
> ```
>     val diskInfos =
>       if (shutdown.get()) {
>         Seq.empty[DiskInfo]
>       } else {
>         storageManager.updateDiskInfos()
>         workerInfo.updateThenGetDiskInfos(storageManager.disksSnapshot().map { disk =>
>           disk.mountPoint -> disk
>         }.toMap.asJava).values().asScala.toSeq
>       }
> ```

Agree

File: client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala [hunk 3]
```diff
@@ -115,12 +115,15 @@ class WorkerStatusTracker(
       val blacklistMsg = blacklist.asScala.map { case (worker, (status, time)) =>
         s"${worker.readableAddress()}   ${status.name()}   $time"
       }.mkString("\n")
+      val shuttingDownMsg = shuttingWorkers.asScala.map { _.readableAddress() }.mkString(";")
```
Comment (pan3793) [inline client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:118]:
```suggestion
      val shuttingDownMsg = shuttingWorkers.asScala.map(_.readableAddress()).mkString("\n")
```

File: client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala (hunk 1 above)
Comment (zhongqiangczq) [reply client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:204]: This function's origin meaning is that only newly added shutdownworkers can be returned (which is not in shuttingWorkers ), but now it returns all the shutdownworkers. Flink's shufflemaster will register ShuffleResourceTracker as workerStatusListeners. ShuffleResourceTracker should juse processes the newly add shutdownworkers. Is that right ? @RexXiong

Comment (RexXiong) [reply client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:204]: ShuffleResourceTracker of Flink doesn't care shutdown workers currently @zhongqiangczq

**Comment by waitinfuture** [comment]
> 

We want a worker be either in the excluded list, or in the shutting down list, exclusively. If a worker is in excluded list, then LIfecycleManager will not trigger commit files when handle heartbeat response; on the other hand, if a worker is in the shutting down list, LifecycleManager will trigger commit files on it. So we must make it correct that a shutting down worker be in the correct list.

**Comment by waitinfuture** [comment]
Hi @pan3793 @RexXiong @zhongqiangczq , I've addressed the comments and refined the logic

File: client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala [hunk 4]
```diff
@@ -53,10 +53,18 @@ class WorkerStatusTracker(
     }
   }
 
-  def isWorkerAvailable(worker: WorkerInfo) = {
+  def workerAvailable(worker: WorkerInfo) = {
```
Comment (pan3793) [inline client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:56]: The function name `isXXX` is fine, it's consistent with others `verbXYZ`
```
isWorkerAvailable
isWorkerOfPartitionExcluded
```

Comment (waitinfuture) [reply client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:56]: I feel it a bit verbose..

**Comment by pan3793** [comment]
> We want a worker be either in the excluded list, or in the shutting down list, exclusively. If a worker is in excluded list, then LIfecycleManager will not trigger commit files when handle heartbeat response; on the other hand, if a worker is in the shutting down list, LifecycleManager will trigger commit files on it. So we must make it correct that a shutting down worker be in the shutting down list.

Thanks for the explanation, it's worth to comment it in the code and adding it to the PR description.

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 5]
```diff
@@ -587,6 +589,9 @@ private boolean revive(
               ClassTag$.MODULE$.apply(PbChangeLocationResponse.class));
       // per partitionKey only serve single PartitionLocation in Client Cache.
       StatusCode respStatus = Utils.toStatusCode(response.getStatus());
+      if (!response.getExcluded()) {
```
Comment (RexXiong) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:592]: Why not remove newLocation worker from blacklist, And IMO client side blacklist would sometimes more timely then LifecycleManager, so it maybe not credible from LM.

Comment (waitinfuture) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:592]: The newLocation should not be in executor's excluded list because revive will exclude the oldPartition. Executor's excluded list will be added in LifecycleManager's excluded list, see ```ChangePartitionManager::handleRequestPartitions```. Currently, LM has a more complex and accurate strategy to mange its excluded list, so I think it's more accurate than executor side.

File: client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala (hunk 4 above)
Comment (waitinfuture) [reply client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:56]: done

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala [hunk 6]
```diff
@@ -226,9 +229,11 @@ class ChangePartitionManager(
...
         }
       }.foreach { case (newLocation, requests) =>
-        requests.map(_.asScala.toList.foreach(_.context.reply(
-          StatusCode.SUCCESS,
-          Option(newLocation))))
+        requests.map(_.asScala.toList.foreach(req =>
+          req.context.reply(
+            StatusCode.SUCCESS,
+            Option(newLocation),
+            lifecycleManager.workerStatusTracker.excluded(req.oldPartition))))
```
[the remark this answers is not in the document]
Comment (waitinfuture) [reply client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala]: done

**Comment by waitinfuture** [comment]
> > We want a worker be either in the excluded list, or in the shutting down list, exclusively. If a worker is in excluded list, then LIfecycleManager will not trigger commit files when handle heartbeat response; on the other hand, if a worker is in the shutting down list, LifecycleManager will trigger commit files on it. So we must make it correct that a shutting down worker be in the shutting down list.
> 
> Thanks for the explanation, it's worth to comment it in the code and adding it to the PR description.

Updated the description. Since I reverted the modification to Master, and instead changes Worker's heartbeat, I don't put the comment there.

**Comment by waitinfuture** [comment]
Hi @RexXiong @pan3793 , I've addressed the comments, please take a look again.

**Review by RexXiong — APPROVED** [review]
LGTM

File: common/src/main/proto/TransportMessages.proto [hunk 7]
```diff
@@ -220,6 +220,7 @@ message PbRevive {
 message PbChangeLocationResponse {
   int32 status = 1;
   PbPartitionLocation location = 2;
+  bool available = 3;
```
Comment (pan3793) [inline common/src/main/proto/TransportMessages.proto:223]: I suppose the 0.2 client just simply ignores it and still works well, right?

**Review by pan3793 — APPROVED** [review]

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala [hunk 8]
```diff
@@ -226,9 +229,12 @@ class ChangePartitionManager(
           location -> Option(requestsMap.remove(location.getId))
         }
       }.foreach { case (newLocation, requests) =>
-        requests.map(_.asScala.toList.foreach(_.context.reply(
-          StatusCode.SUCCESS,
-          Option(newLocation))))
+        requests.foreach(_.asScala.foreach { req =>
```
Comment (pan3793) [inline client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala:232]: why this change is reverted?

File: client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (pan3793) [reply client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala]: I see, `requests.toSeq...` could fix it.

File: common/src/main/proto/TransportMessages.proto (hunk 7 above)
Comment (waitinfuture) [reply common/src/main/proto/TransportMessages.proto:223]: This PR is OK, but I do find compatibility exceptions when testing 0.2.1 client with main server. I will investigate and send new PRs to solve.
